### PR TITLE
Support Checkpointing - Keras WandbCallback

### PR DESCRIPTION
**Pull request by ayulockin**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
I an earlier PR #3276 we added the support to save the subclassed model or model with custom loss/layer/etc. in the `SavedModel`  format as W&B Artifacts. 

In this PR, the idea is to expand on that and allow user to control checkpointing interval and method by introducing two new arguments to the `WandbCallback` - `save_best_only` and `save_model_frequency`.

### Current behaviour:

If `save_model=True`, the `WandbCallback` will try to save the model in `.h5` format and log it as W&B files. (We have a deprecation warning to remove the support for `.h5` in future since `SavedModel` format is the recommended way of saving to and loading from). If it fails to save it in the `.h5` format, it will save the model in the `SavedModel` format everytime, the `monitor` improves. 

### New behaviour:

**If the model is to be saved in the `SavedModel` format.** This PR introduces two new behaviours:

* If `save_best_only=True`, the model will be saved as versions of an artifact when the `monitor` improves. It will also add an alias `best`. Thus while training a model, if it overfits, you will get the best checkpoint. This behavior takes precedence over the next one.

image

* If `save_model_frequency=5` and `epochs=20`, the callback will save 4 versions as model artifact. This is useful if the model is going to converge after many epochs and you want to checkpoint at interval to save storage (other factors). This adds an alias `latest`.

image

### Questions:

* Should we remove the support of `.h5` model saving? If yes, should it be done as a separate PR or this PR itself?

Testing
-------
`yea` test.

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

**Comment by kptkin** [comment]
cc-ing: @vwrj and @tssweeney for visibility and input from model registry perspective

File: wandb/integration/keras/keras.py [hunk 1]
```diff
@@ -1006,19 +1025,46 @@ def _save_model(self, epoch):
...
-        # TODO: Replace this manual artifact creation with the `log_model` method
-        # after `log_model` is released from beta.
+        try:
+            if self.save_best_only:
+                current = logs.get(self.monitor)
+                if current is None:
+                    wandb.termwarn(
+                        f"Can save best model only with {self.monitor}"
+                        "available, skipping."
+                    )
```
Comment (kptkin) [inline wandb/integration/keras/keras.py:1039]: Could you reword this one? I'm not clear what is the warning suppose to mean

File: wandb/integration/keras/keras.py [hunk 2]
```diff
@@ -1006,19 +1025,46 @@ def _save_model(self, epoch):
...
+                        self._save_model_and_log_to_artifact(aliases)
+                        self._art_best = current
+            else:
+                if self.save_model_frequency > 0 and (
+                    self._epochs_since_last_save >= self.save_model_frequency
+                ):
+                    self._epochs_since_last_save = 0
+                    aliases = ["latest", f"epoch_{epoch}"]
+                    self._save_model_and_log_to_artifact(aliases)
+        except Exception as e:
```
Comment (kptkin) [inline wandb/integration/keras/keras.py:1054]: Is there a way to catch more concrete exceptions here?

File: tests/functional_tests/keras/test_keras_checkpointing.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import os
```
Comment (kptkin) [inline tests/functional_tests/keras/test_keras_checkpointing.py:1]: can you test failure cases as well?

File: wandb/integration/keras/keras.py (hunk 1 above)
Comment (ayulockin) [reply wandb/integration/keras/keras.py:1039]: sure.

**Comment by kptkin** [comment]
@ayulockin do you want this PR to be included in the upcoming release?

**Comment by ayulockin** [comment]
Thanks for converting it to draft. I should have done it when we decided not to work on it in the short term. My bad :)

**Comment by ayulockin** [comment]
Closing this in favor of #4201.
